### PR TITLE
feat: add --branch option to clone or checkout different dotfiles branch

### LINKED ISSUE #8282: Allow passing git arguments to coder dotfiles
**Use Case: we want to be able to have a single dotfiles repo, and allow people to create branches to try things out.** 

We are using Coder as a private GItHub App, and therefore it doesn't have access to clone private repos. Since it's a private GitHub app, user's can't install it on their own user. Since it's a GitHub app, it can't clone private repo's, and we don't want all of our dotfiles repo public to the internet. We also don't want user's creating a repo per user / team, as that would be thousands of junk repo's in our org. This leaves us with: what about using branches on top of an existing repo?

To fix this, we would need https://github.com/coder/coder/blob/f0bd258ff1ad8fcc1845ef1c7395ccec232fedcd/cli/dotfiles.go#L19 modified to accept git arguments.

### PR DESCRIPTION
**Pull request by Emyrk**
If `--branch` is specified, the `clone` will clone that branch.

The code will always do a `git branch --show-current` and checkout the specified branch if the flag is provided.

Fixes: https://github.com/coder/coder/issues/8282

# How it works

1. If dotfiles does not exist
   -  If `--branch` is specified, git clone uses `git clone --branch <xxx> ...`
2. If dotfiles does exist
   - Runs `git pull` (
      - Always run this before a checkout because the new branch might not be fetched. So a `git checkout` might fail before a pull.
4. If `--branch` exists
   - Run `git checkout <branch>`
   - Run `git pull --ff-only` (The earlier pull was on the wrong branch)

### CODE REVIEW

File: cli/dotfiles.go [hunk 1]
```diff
@@ -140,6 +146,23 @@ func (r *RootCmd) dotfiles() *clibase.Cmd {
 				_, _ = fmt.Fprintln(inv.Stdout, cliui.DefaultStyles.Error.Render("Failed to update repo, continuing..."))
 			}
 
+			if dotfilesExists && gitbranch != "" {
+				// If the repo exists and the git-branch is specified, we need to check out the branch. We do this after
+				// git pull to make sure the branch was pulled down locally. If we do this before the pull, we could be
+				// trying to checkout a branch that does not yet exist locally and get a git error.
```
Comment (mafredri) [inline cli/dotfiles.go:152]: Git pull will only forward the current branch. The requested branch may be available locally but not forwarded. If we pull main and checkout other, other will still be stale until the next time this runs (judging from this comment, didn't look at the code yet).

File: cli/dotfiles.go [hunk 2]
```diff
@@ -246,11 +269,60 @@ func (r *RootCmd) dotfiles() *clibase.Cmd {
...
+	gitSSHCommand string
+	gitBranch     string
+}
+
+func ensureCorrectGitBranch(baseInv *clibase.Invocation, params ensureCorrectGitBranchParams) error {
+	dotfileCmd := func(cmd string, args ...string) *exec.Cmd {
+		c := exec.CommandContext(baseInv.Context(), cmd, args...)
+		c.Dir = params.repoDir
+		c.Env = append(baseInv.Environ.ToOS(), fmt.Sprintf(`GIT_SSH_COMMAND=%s -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no`, params.gitSSHCommand))
+		var out bytes.Buffer
```
Comment (mafredri) [inline cli/dotfiles.go:295]: Out created multiple times? Perhaps use io.Discard below as a default?

File: cli/dotfiles.go [hunk 3]
```diff
@@ -246,11 +269,60 @@ func (r *RootCmd) dotfiles() *clibase.Cmd {
...
+	var out bytes.Buffer
+	c.Stdout = &out
+	err := c.Run()
+	if err != nil {
+		return xerrors.Errorf("getting current git branch: %w", err)
+	}
+
+	if strings.TrimSpace(out.String()) != params.gitBranch {
+		// Checkout and pull the branch
+		c := dotfileCmd("git", "checkout", params.gitBranch)
```
Comment (mafredri) [inline cli/dotfiles.go:311]: This differs from earlier comment about pulling current branch first?

Also a branch may not have an upstream. Perhaps we should git fetch at some point too?

**Review by mafredri — COMMENTED** [review]
Nice feature! Might've misunderstood some parts but left a few questions.

File: cli/dotfiles.go (hunk 1 above)
Comment (Emyrk) [reply cli/dotfiles.go:152]: Correct, but this code does a `git checkout + git pull` if we are on the wrong branch.

So the first step is either a `git clone` or a `git pull`. 
The second step is a `git checkout + git pull` **if and only if** `git branch --show-current` is not the correct branch.

Tl;dr we run 2 pulls

File: cli/dotfiles.go (hunk 2 above)
Comment (Emyrk) [reply cli/dotfiles.go:295]: Whoops this was a copy/paste mistake. That should be baseInv.StdOut by default.

File: cli/dotfiles.go (hunk 3 above)
Comment (mafredri) [reply cli/dotfiles.go:311]: My bad, looks like `git pull` does a fetch even if there's no upstream. Forget I said anything. 😅

File: cli/dotfiles.go [hunk 4]
```diff
@@ -140,6 +146,23 @@ func (r *RootCmd) dotfiles() *clibase.Cmd {
...
+				_, _ = fmt.Fprintf(inv.Stdout, "Dotfiles git branch %q specified\n", gitbranch)
+				err := ensureCorrectGitBranch(inv, ensureCorrectGitBranchParams{
+					repoDir:       dotfilesDir,
+					gitSSHCommand: gitsshCmd,
+					gitBranch:     gitbranch,
+				})
+				if err != nil {
+					// Do not block on this error, just log it and continue
+					_, _ = fmt.Fprintln(inv.Stdout,
+						cliui.DefaultStyles.Error.Render(fmt.Sprintf("Failed to use branch %q (%s), continuing...", err.Error(), gitbranch)))
```
Comment (mafredri) [inline cli/dotfiles.go:162]: Would it make sense to use a non-zero exit code if this happens? Otherwise the user might expect one thing but receive another due to the silent failure. (Not saying we can't continue from here, but set a `branchFailed = true` and then exit 1 at the end.

**Review by mafredri — APPROVED** [review]
Once the stdout thing is fixed, I see no blockers. LGTM and thanks for answering my questions! 👍🏻

File: cli/dotfiles.go (hunk 4 above)
Comment (Emyrk) [reply cli/dotfiles.go:162]: That is an idea. I am not sure, I was copying the existing behavior.

See this: 
https://github.com/coder/coder/blob/e1d7dc38fcf323114a2b1ab166b71be0d75d43ca/cli/dotfiles.go#L142-L142

If the `git pull --ff-only` fails then we continue, even though we are now operating in a stale dotfiles. I was maintaining that behavior of continuing.

If we decide to exit code 1, we should go back and also change the previous behavior.
